### ISSUE TITLE
Excalidraw does not work with a lot of users

### ISSUE THREAD

**Issue by kolaente** [excalidraw/excalidraw#1808]
First of all, thank you for this awesome project!

We just tried it in a lecture, but with 43 people on one board, users had different amounts of lag between 0-5min each, making the real-time update collaboration pretty much unusable.

When looking at the websocket connection used to make this all possible, it seems like excalidraw opens a connection for each user, resulting in a lot of connections: 

![excalidraw-ws](https://user-images.githubusercontent.com/13721712/85371871-40002e00-b531-11ea-80df-5f80bf178757.gif)

You can see in the gif there were already 21K messages after just one minute, which makes roughly 1 message each 3ms. That seems quite a lot.

To improve this, it may be an idea to let each client have just one connection to a server which then aggregates and distributes the messages.

To me 43 concurrent users does not seem like a lot but I'm not sure if excalidraw has been built with that use case in mind.

**Comment by kbariotis** [comment]
Thank you for raising this @kolaente. Rest assured that we are working on improving performance especially in the collaboration front.

> To improve this, it may be an idea to let each client have just one connection to a server which then aggregates and distributes the messages.

Just to clarify: that's the current state, there is only one connection per client with a centralized server. The real bottleneck is in the frequency/size of the messages and the time/effort it needs for each client to replicate those on their end.

**Comment by dwelle** [comment]
Yes, your GIF shows incoming messages so what you're likely seeing is all those 43 clients sending their mouse/scene data which the server then distributes to all other clients, including you.

We could in theory aggregate messages on the server, but that would require queuing.

**Comment by justshrey** [comment]
Happened to us today as well. We were really not able to use it at all and it was about 7 of us only. For some reason, the browser actually "hung" for most of us. 

I haven't really figured out yet why but i think it could be because

A) We had a fairly complex template setup before every one logged in

B) Some of them were on a relatively laggy connection ( we were all on zoom at the same time while on the board so might have been taking a lot of bandwidth already - the zoom call was working really clearly )

Not sure if firefox has anything to with it but all of us were on Firefox.

**Comment by dwelle** [comment]
What's was the approximate geo location of most of you, btw? Just so that we get more anecdotal evidence for the importance of being close to the data center.

**Comment by kolaente** [comment]
@dwelle I'm not 100% sure but I think most of us were in like a 100km radius, definitely all in the same country (not sure how relevant that is).

**Comment by NMinhNguyen** [comment]
> @dwelle I'm not 100% sure but I think most of us were in like a 100km radius, definitely all in the same country (not sure how relevant that is).

What country? That might help too.

**Comment by kolaente** [comment]
@NMinhNguyen Germany, in the neighbourhood of Frankfurt.

**Comment by dwelle** [comment]
> @NMinhNguyen Germany, in the neighbourhood of Frankfurt.

Yea. We have a collab server in the US (West Coast I think). I live in Europe as well, and the latency to the server affects collab a lot in my experience.

**Comment by frederikfab** [comment]
Same issue here.. we were 14 people, and the tool performed very poorly (some lost their input, object order from back to front flipped around a lot)

**Comment by dwelle** [comment]
> some lost their input

meaning the updates they made were not synced correctly?

> object order from back to front flipped around a lot

Yes, z-index syncing had issues forever, but recent changes to syncing (to actually improve performance) made it way worse where currently z-index is pretty much broken completely. I've updated https://github.com/excalidraw/excalidraw/issues/1376 to reflect this.

**Comment by frederikfab** [comment]
> meaning the updates they made were not synced correctly?
Yes. And people were scared to try to join the room again, as a model dialog pointed out that data could be lost.
 
> > object order from back to front flipped around a lot
> 
> Yes, z-index syncing had issues forever, but recent changes to syncing (to actually improve performance) made it way worse where currently z-index is pretty much broken completely. I've updated #1376 to reflect this.

Cool, thanks. I would be happy if the relative z-indexes would be maintained (here even only within objects that are grouped). It only really matters if the bounding-boxes of the elements overlap right?

**Comment by dwelle** [comment]
> Yes. And people were scared to try to join the room again, as a model dialog pointed out that data could be lost.

Ouch. We'll need to improve that dialog for sure.

> It only really matters if the bounding-boxes of the elements overlap right?

In practice, yeah. But we'll need to fix it anyway, and as you said, it also messes up grouped elements and stuff.
